src/mscorlib/corefx/System/Buffers/DefaultArrayPool.cs

```diff
@@ -125,37 +153,133 @@ public override void Return(T[] array, bool clearArray = false)
             {
                 log.BufferReturned(array.GetHashCode(), array.Length, Id);
             }
         }
+
+        /// <summary>Provides a global bucket for a particular buffer length size.</summary>
+        private sealed class GlobalBucket : MultiProducerMultiConsumerBoundedQueue<T[]>
+        {
+            /// <summary>The length of buffers that should be created for this bucket.</summary>
+            public int BufferLength { get; }
+
+            /// <summary>Creates the pool with numberOfBuffers arrays where each buffer is of bufferLength length.</summary>
+            public GlobalBucket(int bufferLength, int numberOfBuffers) : base(numberOfBuffers)
+            {
+                Debug.Assert(bufferLength > 0, $"Buffer length must be positive, got {bufferLength}");
+                BufferLength = bufferLength;
+            }
+        }
+
+        /// <summary>Provides a TLS bucket for a particular buffer length size.</summary>
+        /// <remarks>This type doesn't actually contain TLS, but rather is intended to be used in TLS.</remarks>
+        private sealed class TlsBucket
+        {
+            /// <summary>The maximum number of arrays it can hold.</summary>
+            private readonly int _capacity;
+            /// <summary>The arrays, lazily initialized on first use.</summary>
+            private WeakReference<T[]>[] _arrays;
+
+            /// <summary>Initializes the stack.</summary>
+            /// <param name="capacity">The maximum number of arrays that can be stored.</param>
+            public TlsBucket(int capacity)
+            {
+                _capacity = capacity;
+            }
+
+            /// <summary>Attempt to push an array onto the stack.</summary>
+            /// <param name="array">The array to push.</param>
+            /// <returns>true if the array was stored; false if the stack was full.</returns>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public bool TryPush(T[] array)
+            {
+                // If we've never pushed on this thread before, initialize the bucket,
+                // including allocating the array and filling it with weak ref objects.
+                // The array being pushed gets stored in the first one.
+                if (_arrays == null)
+                {
+                    _arrays = new WeakReference<T[]>[_capacity];
+                    _arrays[0] = new WeakReference<T[]>(array);
+                    for (int i = 1; i < _arrays.Length; i++)
+                    {
+                        _arrays[i] = new WeakReference<T[]>(null);
+                    }
+                    return true;
+                }
+
+                // This isn't our first time pushing on this thread. Iterate through
+                // the weak references looking for one that doesn't have an array,
+                // and store the pushed array into it.
+                foreach (WeakReference<T[]> wr in _arrays)
```

Comment:
Yeah.  As I noted, what I really wanted was something that would move the buffers from here to the global pool, instead of dropping them.  To my knowledge we just don't have anything like that existing yet.  I'd tried using finalizable objects and resurrection to do it, essentially instead of having a weak ref to an array, having a weak ref to a poolable resurrectable finalizable object wrapping an array, and the finalizer would TryEnqueue the array on to the global queue.  Perf tanked, but it's possible I messed it up somehow... I'll try again now that I've made other changes.